etcd3 store: retry with live object on conflict if there was a suggestion

ncdc opened the pull request:
Retry with a live object instead of the cached version if the watch
cache receives a conflict trying to do the update.

Fixes #41892

```release-note
Fix spurious resource version conflict error during `kubectl replace`.
```

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/43152)
<!-- Reviewable:end -->

ncdc commented:
Was the change to use the watch cache introduced in 1.5 or 1.6? I'm wondering if we need a release note.

ncdc commented:
If we want this for 1.6, we should move #41892 back to the v1.6 milestone too

xiang90 commented on staging/src/k8s.io/apiserver/pkg/storage/cacher_test.go:
```diff
@@ -0,0 +1,92 @@
+	return nil
+}
+
+func TestGuaranteedUpdate(t *testing.T) {
```
Can we give this test with a more descriptive name?

ncdc replied:
@xiang90 TestGuaranteedUpdateUsesLiveObjectOnConflict?

xiang90 replied:
DropSuggestionsOnConflict?

ncdc replied:
Fixed

xiang90 commented:
@ncdc 

I feel it would be cleaner just do a pass through inside storage's GuaranteedUpdate func. But I guess we might argue to keep the actual storage implement simple for potential other backends.

I do not have strong feeling here.
So LGTM.

ncdc commented:
@xiang90 I don't understand what you're suggesting. Could you please clarify?

liggitt commented:
I'm trying to put together a test to satisfy myself that this does the right thing for Updates on AllowCreateOnUpdate items

xiang90 commented:
>  I don't understand what you're suggesting. Could you please clarify?

The storage implementation with curObj here: https://github.com/kubernetes/kubernetes/pull/43152/files#diff-dc17590f6b960d2fea2b44de5c0fbc10R529 

The storage implementation can be smarter about dropping the suggested curObj and try again inside the func. So we do not need to explicitly drop the curObj outside the storage implementation.

ncdc commented:
@xiang90 ok I understand, thanks. I'll wait for more review comments before I make any adjustments.

ncdc commented:
@k8s-bot gce etcd3 e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot kops aws e2e test this
@k8s-bot gci gce e2e test this

xiang90 commented:
@ncdc Thanks. I am actually OK with what you have today. Just a minor suggestion.

ethernetdan commented:
@k8s-bot gce etcd3 e2e test this
@k8s-bot cvm gce e2e test this
@k8s-bot kops aws e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot verify test this

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/storage/cacher.go:
```diff
@@ -526,11 +526,17 @@ func (c *Cacher) GuaranteedUpdate(
+				return err
+			}
+		} else {
+			glog.Errorf("couldn't copy object: %v", copyErr)
```
No! :)

wojtek-t replied:
Sorry - I don't understand this comment.

ncdc replied:
@smarterclayton all I did was move the previously existing line 👅 

@wojtek-t he wants `utilruntime.HandleError`

ncdc replied:
Fixed

liggitt commented on staging/src/k8s.io/apiserver/pkg/storage/cacher.go:
```diff
@@ -526,11 +526,17 @@ func (c *Cacher) GuaranteedUpdate(
-		glog.Errorf("couldn't copy object: %v", copyErr)
 	}
-	// If we couldn't get the object, fallback to no-suggestion.
+	// If we couldn't get the object or we got a conflict trying to update, fallback to no-suggestion.
```
this doubles the cost of a legitimate conflict. I wonder if we should be smarter in the storage impl and only retry on conflict from suggestion if the next live lookup has a different resource version

wojtek-t replied:
That sounds like a good suggestion, though i would be afraid of doing this for 1.6.

ncdc replied:
Added TODO

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/storage/cacher.go:
```diff
@@ -526,11 +526,17 @@ func (c *Cacher) GuaranteedUpdate(
 		currObj, copyErr := c.copier.Copy(elem.(*storeElement).Object)
 		if copyErr == nil {
-			return c.storage.GuaranteedUpdate(ctx, key, ptrToType, ignoreNotFound, preconditions, tryUpdate, currObj)
+			err := c.storage.GuaranteedUpdate(ctx, key, ptrToType, ignoreNotFound, preconditions, tryUpdate, currObj)
```
I'm ok with the structure of this, it's pretty clear what is happening and is lower risk for 1.6.  A TODO  would be good to call out that we may want to consider the alternative approach.

Waiting for Jordan's feedback

wojtek-t replied:
+1

ncdc replied:
Done

ncdc commented:
Comments addressed. PTAL

ncdc commented:
Given that optimization to avoid a get in the watch cache for GuaranteedUpdate was added back in 1.5 (https://github.com/kubernetes/kubernetes/commit/a1090151efaccaa663e077f7dc89202575e754b7), I don't think we should rush this fix in for 1.6, as it's not a post-1.5 regression. We should, however, merge #43192, as that should hopefully kill the flake with a retry.

liggitt commented:
for when we get back to this, I had a test that recreated the conflict error and passed with this PR:
https://gist.github.com/liggitt/e6a8d964d5a0776386c7b4fdc00b6bdd

lavalamp commented:
/assign @jpbetz

lavalamp commented:
/assign @caesarxuchao

liggitt commented:
we ran into a scenario where a watch cache had a persistently stale version of a resource (get with ?resourceVersion=0 shows a stale version of the resource indefinitely). we're still trying to figure out how this occurred, but in a scenario like that, it becomes impossible to issue an update with a resourceVersion without this fix, since the cacher always passes in the stale version, always gets back a conflict error, and always returns with an error.

liggitt commented:
cc @enj

ncdc commented:
Do you want me to resume working on this? From what I recall, we weren't certain we wanted to do it this way?

liggitt commented:
I think so. I do think we have to move the logic into the storage impl where the suggestion is used (currently, only etcd3.store#GuaranteedUpdate uses it).

I'm envisioning something like this:
```go
ret, ttl, err := s.updateState(origState, tryUpdate)
if err != nil {
  // if we were working with potentially stale data
  if mustCheckData && IsConflict(err) {
    // actually fetch
    getResp, err := s.client.KV.Get(ctx, key, s.getOps...)
    if err != nil {
      return err
    }
    origState, err = s.getState(getResp, key, v, ignoreNotFound)
    if err != nil {
      return err
    }
    mustCheckData = false
    // retry
    continue
  }

  return err
}
```

ncdc commented:
Ok, I can see about getting something tomorrow. Otherwise it'll probably
have to wait a week, or someone else can pick it up next week while I'm
mostly unavailable.

On Thu, Sep 14, 2017 at 3:37 PM, Jordan Liggitt <notifications@github.com>
wrote:

> I think so. I do think we have to move the logic into the storage impl
> where the suggestion is used (currently, only etcd3.store#GuaranteedUpdate
> uses it).
>
> I'm envisioning something like this:
>
> ret, ttl, err := s.updateState(origState, tryUpdate)if err != nil {
>   // if we were working with potentially stale data
>   if mustCheckData && IsConflict(err) {
>     // actually fetch
>     getResp, err := s.client.KV.Get(ctx, key, s.getOps...)
>     if err != nil {
>       return err
>     }
>     origState, err = s.getState(getResp, key, v, ignoreNotFound)
>     if err != nil {
>       return err
>     }
>     mustCheckData = false
>     // retry
>     continue
>   }
>
>   return err
> }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/43152#issuecomment-329586827>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYhP05EVYRFTeHK6a-phlj8ERkNzsks5siYCOgaJpZM4MeVj7>
> .
>

wojtek-t commented:
I like @liggitt suggestion - that should solve this particular problem that you observed.

I looked into the code deeper and so far I have completely no idea how it can happen that one of the objects is in stale state, but watch events are delivered. I will continue looking and get back if I find something.

ncdc committed:
etcd3 store: retry w/live object on conflict

In GuaranteedUpdate, if it was called with a suggestion (e.g. via the
watch cache), and the suggested object is stale, perform a live lookup
and then retry the update.

Signed-off-by: Andy Goldstein <andy.goldstein@gmail.com>

ncdc commented:
@liggitt @wojtek-t @smarterclayton updated, ptal

enj commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -299,6 +300,22 @@ func (s *store) GuaranteedUpdate(
 		ret, ttl, err := s.updateState(origState, tryUpdate)
 		if err != nil {
+			// It's possible we were working with stale data
+			if mustCheckData && apierrors.IsConflict(err) {
```
Lets make this logic a function since its the same as above https://github.com/kubernetes/kubernetes/pull/43152/files#diff-0f11327dd3daafdfe8e326e895616f66R284

ncdc replied:
I'm not sure that's entirely necessary given that it's only used in 2 places. If it's 3+, definitely... but if you want it, I can do it. Have a suggestion for a name?

ncdc replied:
It's going to end up having to be

```
func (s *store) yourNameHere(ctx context.Contect, key string, v reflect.Value, ignoreNotFound bool) (*objState, error)
```

Not convinced we need it.

ncdc replied:
Ha ok I do see it used in 3 places

liggitt commented:
@wojtek-t we saw the stale cache issue on an etcd cluster that went through some issues with nodes dropping out and rejoining... I'm suspecting it triggered https://github.com/coreos/etcd/issues/8411 where a rejoined etcd node doesn't notify watchers of events that came in as part of the restored snapshot

ncdc commented:
@enj @liggitt updated

enj commented:
LGTM

wojtek-t commented:
LGTM

I will let others also take a look, so just approving for now.

/approve no-issue

liggitt commented:
/lgtm

enj commented:
/lgtm

wojtek-t commented:
@ncdc @liggitt - what do you think about cherrypicking it? I think we should.

liggitt commented:
> @ncdc @liggitt - what do you think about cherrypicking it? I think we should.

absolutely, especially given the etcd bug. I'd like to see this in 1.7 and 1.6

liggitt commented:
separately, it seems like we'd want to pick up an etcd level with a fix for that bug

liggitt commented:
/retest

liggitt commented:
/retest

liggitt commented:
will prep a pick to 1.7 (along with https://github.com/kubernetes/kubernetes/pull/48394 which was never picked as well)
